### ISSUE TITLE
Why is HashMap::iter considered unsafe?

### ISSUE THREAD

**Issue by willfindlay** [aya-rs/aya#82]
HashMap::iter is marked as an unsafe method, and it's not immediately clear why. If it is truly unsafe, there should be a safety contract that outlines how the caller should avoid UB. The same applies to other map types.

**Comment by alessandrod** [comment]
I made it unsafe since the underlying API is fundamentally racy. Eg you can have:

```
map.insert(key, value);
for item in map.iter() {
   /// the item with (key, value) might not be returned here if it's removed by another process
}
```

I had the above case happen to me and it took me a while to realise what was going on. Typically when I have obscure bugs, the first thing I do is look for `unsafe` around the areas where I think the bug might be.

So it is unsafe as in "this might be a foot gun" not  as in "watch out you can trigger UB". That said as I'm writing this I'm realising that probably a bunch of APIs that do syscalls and interact with the OS have this kind of behaviour, so maybe we should just document the possible foot gun and make it safe. Not sure.

**Comment by willfindlay** [comment]
> So it is unsafe as in "this might be a foot gun" not  as in "watch out you can trigger UB". That said as I'm writing this I'm realising that probably a bunch of APIs that do syscalls and interact with the OS have this kind of behaviour, so maybe we should just document the possible foot gun and make it safe. Not sure.

I think that's probably the better approach. At least it feels more idiomatic to me. With a clear warning in the documentation we should be fine imo.

**Comment by willfindlay** [comment]
This was closed in #132
